### ISSUE TITLE
Cannot upsert anything under a node that is going to be related in upsertGraph.

### ISSUE THREAD

**Issue by muffinsparticus** [Vincit/objection.js#782]
Greetings!

I've encountered an issue that's blocking me from moving forward with objection.js at the moment. 

I have a DataModel (A) that has a many to many relationship with another DataModel (B) that has a many to many relation to another DataModel (C). Whenever I perform an upsert with using DataModel A objection always tries to insert the relation between BxC into the proper join table when the relation exists. 

Is this a bug with the upsert graph and if not is there a way to prevent objection from inserting duplicate relations? It seems to handles the AxB relation just fine.

P.S. It also doesn't seem to properly unrelate the BxC the relation when it doesn't exist.

**Comment by koskimas** [comment]
Hi,

unfortunately I have no idea what you mean by that descriptioin. `upsertGraph` is so complex, that your description can mean a hundred different things. Could you use the `reproduction-template.js` file in the repo root and create a simple example of the case you are talking about?

**Comment by koskimas** [comment]
Also could you try the newest 1.0 RC version if you haven't already. It has a bunch of upsertGraph fixes.

**Comment by muffinsparticus** [comment]
Thank you very much for the response! 

The project is breaking when I install version 1.0.0-RC12 and I can confirm it works with version 0.9.4. I'm currently using gulp to pipe the data models through babel to compile it to ES5 javascript. Yet when I try to run the code, after compiling, it gives me the following error message:

{ errorMessage: 'Unexpected token {',
      errorType: 'SyntaxError',
      stackTrace: 
       [ 'Module._compile (module.js:373:25)',
         'Object.Module._extensions..js (module.js:416:10)',
         'Module.load (module.js:343:32)',
         'Function.Module._load (module.js:300:12)',
         'Module.require (module.js:353:17)',
         'require (internal/module.js:12:17)',
         'Object.<anonymous> ( main.js:9:18 )',
         'Module._compile (module.js:409:26)',
         'Object.Module._extensions..js (module.js:416:10)' ] }

Are there any new requirements with version 1.0.0-RC12 that I should be aware of? Ex: new babel dependencies to install. I can provide more info about the build process if necessary.

In the meantime I'll try to get a test example ready using the repo-template.

**Comment by koskimas** [comment]
There's nothing new that should or could break babel.

**Comment by muffinsparticus** [comment]
Thank you for the reply!

I've attached a copy of an example using the repo file as requested below: 
[reproduction-template.js.zip](https://github.com/Vincit/objection.js/files/1704686/reproduction-template.js.zip)

**Comment by koskimas** [comment]
Thanks for the reproduction. There's a known limitation currently that doesn't allow you to relate two nested levels in a single query. This can be really difficult to fix, but I'll try to at least detect the situation and throw a good error if I cannot fix this.

**Comment by muffinsparticus** [comment]
That's fantastic to know! Thank you for sharing that information!

Would you have any recommendations on getting around this limitation or an alternate way of representing the data to achieve a similar parent-child-child tree effect with objection?

**Comment by koskimas** [comment]
I can't think of one. In this particular case, you need to use two separate operations:

```js
  await Person.query().upsertGraph({
    id  : 2,
    name: "person2",
    movies: [ {
      id: 1,
      name: "movie1"
    } ]
  }, {
    relate: true,
    unrelate: true,
  });

  await Movie.query().debug(true).upsertGraph({
    id  : 1,
    name: "movie1",
    animals: [ {
      id: 1,
      name: "animal1"
    } ]
  }, {
    relate: true,
    unrelate: true,
  });
```

But let's see what I can do about this issue.

**Comment by koskimas** [comment]
Note that you can `insert` and `update` any number of nested relations. The `relate` option is what breaks things here.

**Comment by koskimas** [comment]
Hmm.. Your case here is kind of weird. Are you sure you want to do what the reproduction is doing? What it is trying to do is:

1. relate movie1 to person2 by inserting a row `{ movieId: 1, personId: 2  }` to `Person_Movie` table.
2. relate movie1 to animal1 by inserting a row `{ movieId: 1, animalId: 1 }` to `Movie_Animal` table.

And nothing else. Isn't it kind of weird that, while updating `person1` you end up modifying `movie1`'s other relations as well. If some other `person` has a reference to `movie1` it will have the same related animals.

This seems like a pretty weird corner-case. Could you elaborate the actual use case here?

**Comment by koskimas** [comment]
I added an error for this case for now. Fixing this would probably introduce too many corner case bugs and I don't want to break things now since we are about to release 1.0 soon. Fixing this won't be a breaking change and can be added without major version bump at any time. I'll fix this at some point in the future.

**Comment by muffinsparticus** [comment]
Apologies for the delay in the reply!

I attempted your suggestion to separate the upsert graph into many smaller ones and it worked! I generated a separate upsert for each major object in the hierarchy. Then for each object I upsert I strip the corresponding child object down to their ID, and required extra field where necessary, to avoid the many-to-many layer limitation. I can provide more information if needed. 

In regards to the use-case for the example I gave: I'm trying to support put operations on a rest API where I have the entire graph.

**Comment by johannessjoberg** [comment]
Is there any progress with this issue?

**Comment by koskimas** [comment]
Nope

**Comment by johannessjoberg** [comment]
All righty. Are there any plans to deal with this in the near future? If not, is help wanted? :)

**Comment by johannessjoberg** [comment]
I've been thinking of this issue during the weekend and I would not call this a corner case. From reading the documentation, I would assume relations are handled in the same manner, irregardless of nested deep and if child of a node that is going to be related. I understand this issue is only related(pun intended) to the `relate` option, but for me this it's the most usable option of `upsertGraph`.

As always, great job with objection @koskimas ! Super glad to see the first major is in place 👍

**Comment by koskimas** [comment]
I'd merge a PR that fixes this without complicating the code too much. If fixing this requires making the upsertGraph code unreadable (and more error prone), I don't think this should be added.

**Comment by koskimas** [comment]
I @johannessjoberg or someone else attempts to write a PR, here are the mandatory things it must have:

1. All current tests must pass (this should be obvious)
2. The new feature must be tested
3. Must support all current options with boolean **and** relation path values. Especially remember to consider the case where relation paths point to models that are under the model that is about to be related.

Probably the only way to implement this is to recursively call `upsertGraph` for the related nodes, which can be very inefficient in some cases.

**Comment by koskimas** [comment]
Implemented by @johannessjoberg in #920
